structured_sources: Allow build_tgt to be added as source

sp1ritCS opened the pull request:
In some cases, there is a need to use a built library as a "source" for another target. Specifically Java JAR might want to contain their own native JNI libraries in order to provide a single file "executable".

I'm kind of surprised this works out of the box without any further changes 🙃

sp1ritCS committed:
structured_sources: Allow build_tgt to be added as source

In some cases, there is a need to use a built library as a "source" for
another target. Specifically Java JAR might want to contain their own
native JNI libraries in order to provide a single file "executable".

tristan957 commented:
Can you please extend a test?

sp1ritCS commented:
@tristan957 added

tristan957 commented on test cases/java/11 shared library resource/src/com/mesonbuild/SharedLibraryResource.java:
```diff
@@ -0,0 +1,41 @@
+    static {
+        String fullname = System.mapLibraryName(libname);
+        // this is somewhat hacky, but attempt to split off the extension
+        int ext = fullname.indexOf(".");
```
Java doesn't have a stem function for this?

sp1ritCS replied:
It seemingly doesn't. And even if it did, I'd consider a stem function is just as hacky. Ideally you could just access lib prefix and suffix from java, but you can't, so you have to gain access to the suffix for `mkstemp` somehow.

No idea if indexOf or lastIndexOf is better, the first breaks if the library prefix or libname contains a `.`, the second breaks if the extension were to contain a `.` (dunno, imagine a system capable of loading `.so.gz` or so)

tristan957 commented on test cases/java/11 shared library resource/meson.build:
```diff
@@ -0,0 +1,17 @@
+    '-jar',
+    shared_library_resource_jar ,
+  ],
+  protocol : 'exitcode'
```
be consistent with spaces before and after the colon.

tristan957 reviewed (commented):
For skipping on < Java 9, do this:

```meson
jni_dep = dependency('jni')
if jni_dep.version().version_compare('< 1.9')
  error('MESON_SKIP_TEST: this test requires Java 9')
endif
```

You may also want to take a look at the CI images to make sure they aren't set to install Java 8. You'll also want to skip the test if the dependency isn't found.

tristan957 commented:
I don't think you need a whole new test for this. I would extend this one: https://github.com/mesonbuild/meson/tree/master/test%20cases/java/9%20jni. What do you think?

tristan957 commented:
@dcbaker is there any other place that would be a good test in addition to this? Not sure if there is a more generic `structured_sources` test.

sp1ritCS commented:
> I don't think you need a whole new test for this. I would extend this one: https://github.com/mesonbuild/meson/tree/master/test%20cases/java/9%20jni. What do you think?

I've thought about it, but it doesn't really work well with how the test is structured in `src/` and `lib/`. As first we need to generate the native headers in the `src/` tree, then compile the library (`lib/`) that is then again needed in `src/` for the `structured_sources` call.

tristan957 commented:
If you see a good way to refactor the existing test, that would be fine.

sp1ritCS committed:
ciimage/arch: don't force java to be 1.8

Parent introduced unit test (9 jni) that depends on at least java 9,
thus causing the test to be skipped.

tristan957 commented:
We need to get the image jobs to pass. Not your fault. I'm working on fixing Fedora.

tristan957 commented:
If you rebase on master, you should pick up the Fedora CI fixes. I think msys2 will also start to pass.

sp1ritCS commented:
Status report:
- Arch is failing due to mirror issues
- openSUSE & Ubuntu have issues with "dub" dependencies (no idea what that is or how to fix it)
- Fedora has issues with zlib

eli-schwartz commented on ci/ciimage/arch/install.sh:
```diff
@@ -15,6 +15,7 @@ pkgs=(
   doxygen vulkan-validation-layers openssh mercurial gtk-sharp-2 qt5-tools
   libwmf cmake netcdf-fortran openmpi nasm gnustep-base gettext
   python-lxml hotdoc rust-bindgen qt6-base qt6-tools wayland wayland-protocols
+  debugedit glib2 glib2-devel
```
Disable debug support in AUR packages, instead. We are not e.g. testing that valgrind can perform valgrind on scalapack / wxwidgets-gtk2 so there's no reason to spend longer compiling and produce bigger docker containers.

eli-schwartz commented on ci/ciimage/opensuse/install.sh:
```diff
@@ -18,6 +18,7 @@ pkgs=(
   boost-devel libboost_date_time-devel libboost_filesystem-devel libboost_locale-devel libboost_system-devel
   libboost_test-devel libboost_log-devel libboost_regex-devel
   libboost_python3-devel libboost_regex-devel
+  perl-TimeDate # drop this once obs#1187867 made it into factory
```
Since you're not trying to update the opensuse image today, only the Arch one, I would suggest letting opensuse fix this bug and having meson pick that fix up next week on the weekly refresh job.

You're not going to fix the opensuse image anyway unless you get dub to work, but that's tracked by #12143

eli-schwartz replied:
Thanks for debugging the upstream opensuse issue and submitting a fix to them, though. :)

sp1ritCS replied:
coincidentally im probably the one at fault, given that I submitted lcov 2.0 in the first place 🙃

tristan957 commented:
I have added a PR to really fix the fedora test at #13436

sp1ritCS committed:
ciimage/arch: attempt to fix Arch CI

sp1ritCS committed:
testcases/framework/libgcrypt: dont fail with missing libgcrypt-config

libgcrypt 1.11.0 on Arch doesnt ship with libgcrypt-config anymore. It
only provides libgcrypt.pc.

bonzini commented:
@sp1ritCS if you are still interested can you rebase?

bonzini commented:
There are two problems here.

First, this allows passing build targets also to sources for other build targets, which messes up the type checking.

Second, and actually worse, is that this will package the build target prior to installation. This would have the effect of skipping rpath fixing for example, leaving the build path in the artifacts.

So, it seems that this is a bit of a dead end?

sp1ritCS commented:
> Second, and actually worse, is that this will package the build target prior to installation. This would have the effect of skipping rpath fixing for example, leaving the build path in the artifacts.

Indeed. Its the same issue preventing me from adding native library support to #15598
